### ISSUE TITLE
[tt-train] Sequence parallelism needs matmul+collective fusion on the sequence dim

### ISSUE THREAD

**Issue by imichalakTT** [tenstorrent/tt-metal#52944]
Requirements issue for the fused matmul+collective support SP needs. SP is not implemented in
tt-train yet; there is a prototype on the unlanded `imichalak/llama-sp/5-seq-parallelism` branch.

Under SP the residual stream stays sequence-sharded across the TP axis, so every TP linear needs
an `all_gather` on the sequence before it (column-parallel) or a `reduce_scatter` on the sequence
after it (row-parallel), plus the mirror in backward. Four linears per block, so eight collectives
per block per microbatch, each one adjacent to a matmul.

SP therefore needs a fused matmul+collective that moves **dim 2**. The seven fused
matmul+collective ops under `ttnn/cpp/ttnn/operations/experimental/ccl` all move the tensor's last
dim instead, which is the matmul's `K` or `N`.

## What we need

Two ops on `dim=2` of `(B, 1, S, H)`, with an explicit `cluster_axis`. `T` is the mesh extent
along that axis, and `/T` marks a dim this rank holds only a slice of. Gather-then-multiply always
has a whole `K` and a sharded `N`, multiply-then-scatter the reverse, for both of each op's call
sites. Weights are shown `[out, in]`, i.e. with `transpose_b` set; the two dgrad sites consume the
same stored tensor untransposed, so their operand reads in the opposite order.

**Gather-then-multiply** — column-parallel forward, row-parallel backward dgrad:

```
in :  act    [B, 1, S/T, K]    seq-sharded on cluster_axis
      weight [1, 1, N/T, K]
out:  mm     [B, 1, S,   N/T]  full sequence, gathered over cluster_axis
      act    [B, 1, S,   K]    the gathered activation, needed for wgrad
```

**Multiply-then-scatter** — row-parallel forward, column-parallel backward dgrad:

```
in :  act    [B, 1, S,   K/T]
      weight [1, 1, N,   K/T]
out:  rs     [B, 1, S/T, N]    per-rank partials summed over cluster_axis, then split on seq
```

Requirements:

- Linear topology, and Ring.
- `B > 1`.
- A settable `transpose_b`: weights are stored `[out, in]`, so forward needs it set and the
  input-gradient matmul needs it clear.
- Multiply-then-scatter must not materialise the full `[B, 1, S, N]` partial: 16 MB/rank at
  Llama-8B `seq=2048`, four times per block.
- No caller-owned persistent buffers (our autograd allocates per op).
- No per-shape tuning table — a training run sweeps seq lengths across four differently-shaped
  linears per block. Derivable config, or a documented good-enough default.
- Optional: `fuse_swiglu` on gather-then-multiply, folding the block's largest matmul.

## Why the existing ops don't fit

| Op | CCL dim | Also blocking |
|---|---|---|
| `all_gather_matmul_async` | `dim==3` | no `cluster_axis`; asserts batch dims == 1 |
| `all_gather_minimal_matmul_async` | `dim==-1` | gathers `K`, not the sequence; no transposed weight |
| `strided_all_gather_minimal_matmul_async` | `dim==3` | asserts batch dims == 1 |
| `llama_all_gather_matmul_async` | last dim (shape math ignores `dim`) | fixed core grid and a hardcoded shard-shape constant; decode-shaped |
| `matmul_reduce_scatter_async` | `dim==3` | no `cluster_axis`; caller-owned persistent buffers |
| `minimal_matmul_strided_reduce_scatter_async` | `dim==3` | Ring only; asserts batch dims == 1 |
| `llama_rs_matmul` | `dim==3` | not producer→consumer (its RS input is a separate tensor); requires `cluster_axis == 1`, a ring of 2 or 4, and shard height 32 |

`all_gather_matmul_async` and `matmul_reduce_scatter_async` take no `cluster_axis` at all and ring
over every device in the mesh. On a TP-only mesh that coincides with the TP axis, so they appear
to work; on a DP×TP mesh they would also communicate across DP.

## Where the work is

The two sides are not equally far off, and the difference is below the op level in both cases.

The shared reduce-scatter program already handles this case. Both its ring and line
implementations partition `normalized_dim` 0 through 3, with `slice_Ht /= ring_size` for the
sequence dim, and each excludes matmul fusion for exactly one dim:

```cpp
TT_FATAL(!(fuse_op && normalized_dim == 0),
         "reduce_scatter_minimal_async ring implementation can't be fused with matmul when scattering on dim 0");
```

Scattering on dim 2 with a matmul fused is therefore already expressible, on Ring and Linear
both. What stands between that and multiply-then-scatter is `matmul_reduce_scatter_async`'s own
`dim == 3` check, its fixed `cluster_axis`, and its persistent buffers.

The all-gather side has two obstacles instead, and neither is in an op's validation.

The fused all-gather path rejects Linear topology outright:

```cpp
TT_FATAL(!((topology == ccl::Topology::Linear) && fuse_op),
         "linear is not support when using fused for all-gather");
```

And `all_gather_matmul_async` tells the matmul which gathered data is ready through a signaler
carrying slice width and page offsets along the width. There is no field for a row offset.

## Cost of leaving it unfused

Blackhole QuietBox (4× p150), `tp=4`, Llama-8B `seq=2048`, bf16, trace-replayed, 4 links, using
the MGD and fabric tt-train itself selects: `bh_qb_1_4_ring_ring.textproto` →
`FABRIC_2D_TORUS_XY` → Ring, and a line MGD → `FABRIC_2D` → Linear.

The four matmuls in a block cost 1901 µs forward (285 + 202 + 907 + 507). Backward repeats the
four collectives and needs both an input and a weight gradient, so matmul is roughly 3 × 1901 =
5703 µs per block. Against that:

| Topology | Collective per block, fwd + bwd | Share of linear-layer time |
|---|---|---|
| **Linear** (the default) | 1534 µs | **21%** |
| Ring | 1041 µs | 15% |

At 1 link instead of 4 the same accounting gives 46% (Linear) and 34% (Ring).

Treat these as an order of magnitude. The collectives and forward matmuls are measured; the
backward matmul is assumed at twice forward, and the figure is composed arithmetically from
isolated ops rather than read off a profile of a real step. It is a fraction of linear-layer time,
not of step time, which we have not measured.

Two results that shape the ask:

- The sequence dim costs the same as the feature dim. Worst spread is 1.2% on the all-gather, and
  on the Linear 4-link reduce-scatter the sequence dim is 8% *faster*. The `dim == 3` restrictions
  are a software limit, not hardware.
- `all_gather` + `reduce_scatter` ≈ `all_reduce`. SP adds no traffic; it splits the same traffic
  into two halves that each sit next to a matmul, which is what makes fusion possible. It also
  means SP's own win is activation memory, not communication.

Linear costs 1.3–1.7× Ring for the same collective, consistent with a line needing up to `N-1`
hops where a ring needs about `N/2`.

## Prior art

`ring_attention_all_gather_async` already gathers an arbitrary dim and is used on the sequence dim
in `minimax_m3` and tt-dit LTX attention, so sequence-dim addressing over a `cluster_axis` is
established. The closest fused precedent is `minimax_m3`'s attention o_proj (fused matmul +
reduce-scatter on the last dim) — same producer→consumer overlap shape, one axis over.

<details>
<summary>Measured numbers</summary>

4× p150, mesh `(1, 4)`, `cluster_axis=1`, bf16, TILE, trace-captured. Dispatch floor 3.5–6.4 µs
(`[32,32] @ [32,32]`), so these are device times. Both MGDs declare `channels { count: 4 }`.

Ring — `bh_qb_1_4_ring_ring.textproto`, `FABRIC_2D_TORUS_XY`:

| Collective | 1 link | 2 | 3 | 4 |
|---|---|---|---|---|
| `all_gather` dim 2 | 350.5 | 185.2 | 131.0 | **105.9** |
| `reduce_scatter` dim 2 | 395.3 | 225.6 | 161.1 | **154.4** |
| `all_gather` dim 3 | 350.9 | | | 107.2 |
| `reduce_scatter` dim 3 | 394.7 | | | 153.9 |
| `all_reduce` (classic TP) | 743.5 | | | 257.8 |

Linear — same box with a line MGD, `FABRIC_2D`:

| Collective | 1 link | 2 | 3 | 4 |
|---|---|---|---|---|
| `all_gather` dim 2 | 584.3 | 322.5 | 229.9 | **182.9** |
| `reduce_scatter` dim 2 | 617.9 | 334.4 | 242.2 | **200.6** |
| `all_gather` dim 3 | 584.5 | | | 182.6 |
| `reduce_scatter` dim 3 | 618.4 | | | 218.1 |
| `all_reduce` (classic TP) | 1199.8 | | | 394.8 |

Matmuls (`tp=4`, single device; the two runs agree within 1%, values averaged):

| Matmul | Shape | µs |
|---|---|---|
| qkv | `[2048,4096] @ [4096,1536]` | 285.0 |
| out\_proj | `[2048,1024] @ [1024,4096]` | 201.5 |
| gate\_up | `[2048,4096] @ [4096,7168]` | 907.2 |
| w2 | `[2048,3584] @ [3584,4096]` | 507.2 |

All collectives are `dim=2`/`dim=3` on `cluster_axis=1` of a 1×4 mesh, matching what SP would
issue. Activation `[1,1,2048,4096]`, per-rank shard `[1,1,512,4096]`.

</details>

<details>
<summary>Benchmark script</summary>

```
TT_MESH_GRAPH_DESC_PATH=tt-train/configs/mgd/bh_qb_1_4_ring_ring.textproto \
  pytest <path> -k torus_ring -s
TT_MESH_GRAPH_DESC_PATH=<same MGD with dim_types LINE, LINE> \
  pytest <path> -k mesh_linear -s
```

```python
# SPDX-License-Identifier: Apache-2.0
"""Times the collectives and matmuls a Megatron-SP Llama block would issue.

Runs under both fabrics tt-train can select: FABRIC_2D_TORUS_XY (topology Ring) and FABRIC_2D
(topology Linear). tt-train derives these from the MGD dim_types and never uses the 1D fabrics,
so measuring under FABRIC_1D_RING gives optimistic numbers.

Shapes are Llama-3-8B, seq=2048, batch=1, tp=4. Timing only -- no correctness check.
"""
import os
import time

import pytest
import torch
from loguru import logger

import ttnn

HIDDEN = 4096
INTERMEDIATE = 14336
SEQ = 2048
QKV = (32 + 2 * 8) * 128  # 6144

# Pinned to a 1x4 mesh with tp on axis 1: TP, the mesh_device param, mesh_shape and cluster_axis
# below all have to move together.
TP = 4

REPS = 10  # ops per captured trace
ITERS = 20  # trace replays
TRACE = 90000000  # trace region bytes


def _sems(mesh_device, cores, n):
    return [ttnn.create_global_semaphore(mesh_device, cores, 0) for _ in range(n)]


@pytest.mark.parametrize(
    "device_params",
    [
        {"fabric_config": ttnn.FabricConfig.FABRIC_2D_TORUS_XY, "trace_region_size": TRACE},
        {"fabric_config": ttnn.FabricConfig.FABRIC_2D, "trace_region_size": TRACE},
    ],
    ids=["torus_ring", "mesh_linear"],
    indirect=True,
)
@pytest.mark.parametrize("mesh_device", [(1, 4)], indirect=True)
def test_verify(mesh_device, device_params):
    fabric = device_params["fabric_config"]
    torus = "TORUS" in str(fabric)
    topology = ttnn.Topology.Ring if torus else ttnn.Topology.Linear
    name = "torus_ring" if torus else "mesh_linear"
    logger.info(f"=== {name}: fabric={fabric}, topology={topology}, MGD={os.getenv('TT_MESH_GRAPH_DESC_PATH')} ===")

    grid = mesh_device.compute_with_storage_grid_size()
    crs = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(grid.x - 1, grid.y - 1))})
    ag_sem = _sems(mesh_device, crs, 2)
    rs_sem = _sems(mesh_device, crs, 3)
    ar_sem = (_sems(mesh_device, crs, 2), _sems(mesh_device, crs, 3), _sems(mesh_device, crs, 2))
    barrier = ttnn.create_global_semaphore(mesh_device, crs, 0)

    def mk(shape, shard_dim=None):
        t = torch.randn(shape, dtype=torch.bfloat16)
        mapper = (
            ttnn.ShardTensor2dMesh(mesh_device, mesh_shape=(1, 4), dims=(None, shard_dim))
            if shard_dim is not None
            else ttnn.ReplicateTensorToMesh(mesh_device)
        )
        return ttnn.from_torch(t, device=mesh_device, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, mesh_mapper=mapper)

    def timed(fn):
        for _ in range(2):
            fn()
        ttnn.synchronize_device(mesh_device)
        tid = ttnn.begin_trace_capture(mesh_device, cq_id=0)
        for _ in range(REPS):
            fn()
        ttnn.end_trace_capture(mesh_device, tid, cq_id=0)
        ttnn.synchronize_device(mesh_device)
        ttnn.execute_trace(mesh_device, tid, cq_id=0, blocking=False)
        ttnn.synchronize_device(mesh_device)
        t0 = time.perf_counter()
        for _ in range(ITERS):
            ttnn.execute_trace(mesh_device, tid, cq_id=0, blocking=False)
        ttnn.synchronize_device(mesh_device)
        dt = (time.perf_counter() - t0) / (ITERS * REPS) * 1e6
        ttnn.release_trace(mesh_device, tid)
        return dt

    results = {}
    ag_in2 = mk([1, 1, SEQ, HIDDEN], shard_dim=2)
    ag_in3 = mk([1, 1, SEQ, HIDDEN], shard_dim=3)
    rs_in = mk([1, 1, SEQ, HIDDEN])

    for links in (1, 2, 3, 4):
        results[f"all_gather(dim=2) links={links}"] = timed(
            lambda l=links: ttnn.experimental.all_gather_async(
                ag_in2, dim=2, multi_device_global_semaphore=ag_sem, num_links=l,
                topology=topology, cluster_axis=1, barrier_semaphore=barrier,
            )
        )
        results[f"reduce_scatter(dim=2) links={links}"] = timed(
            lambda l=links: ttnn.experimental.reduce_scatter_minimal_async(
                rs_in, dim=2, multi_device_global_semaphore=rs_sem, barrier_semaphore=barrier,
                num_links=l, topology=topology, cluster_axis=1,
            )
        )

    for links in (1, 4):
        results[f"all_gather(dim=3) links={links}"] = timed(
            lambda l=links: ttnn.experimental.all_gather_async(
                ag_in3, dim=3, multi_device_global_semaphore=ag_sem, num_links=l,
                topology=topology, cluster_axis=1, barrier_semaphore=barrier,
            )
        )
        results[f"reduce_scatter(dim=3) links={links}"] = timed(
            lambda l=links: ttnn.experimental.reduce_scatter_minimal_async(
                rs_in, dim=3, multi_device_global_semaphore=rs_sem, barrier_semaphore=barrier,
                num_links=l, topology=topology, cluster_axis=1,
            )
        )
        results[f"all_reduce links={links} (classic TP)"] = timed(
            lambda l=links: ttnn.experimental.all_reduce_async(
                rs_in, 1, mesh_device, ar_sem[0], ar_sem[1], ar_sem[2],
                ttnn.ReduceType.Sum, topology=topology, num_links=l,
            )
        )

    def bench_mm(label, m, k, n):
        a, w = mk([1, 1, m, k]), mk([1, 1, n, k])
        results[label] = timed(lambda: ttnn.linear(a, w, transpose_b=True))

    bench_mm(f"mm qkv      [{SEQ},{HIDDEN}]@[{HIDDEN},{QKV // TP}]", SEQ, HIDDEN, QKV // TP)
    bench_mm(f"mm out_proj [{SEQ},{HIDDEN // TP}]@[{HIDDEN // TP},{HIDDEN}]", SEQ, HIDDEN // TP, HIDDEN)
    bench_mm(f"mm gate_up  [{SEQ},{HIDDEN}]@[{HIDDEN},{2 * INTERMEDIATE // TP}]", SEQ, HIDDEN, 2 * INTERMEDIATE // TP)
    bench_mm(f"mm w2       [{SEQ},{INTERMEDIATE // TP}]@[{INTERMEDIATE // TP},{HIDDEN}]", SEQ, INTERMEDIATE // TP, HIDDEN)
    bench_mm("mm tiny     [32,32]@[32,32] (dispatch floor)", 32, 32, 32)

    logger.info(f"##### {name} RESULTS (us/op, traced {REPS}x{ITERS}) #####")
    for k, v in results.items():
        logger.info(f"[{name}] {k:52s} {v:9.1f}")
```

</details>
